Token Registrations and Defi Models

mrstepanovic opened the pull request:
## Summary
This PR enables 2 new categories of analytics:
1. Token registrations
2. Defi for Arcane Finance (swaps and liquidity pool actions)

It also modifies smaller aspects of the codebase:
* The overview markdown file
* The location of recency tests, moved from `/gold/core` to `/gold`

### What to focus on in your PR review
* The schemas produced in `fact__swaps`, `fact__liquidity_pool_actions`, and `dim__token_registrations`
* Test design (recency and rigor across silver + gold models)
* Tagging
* Clustering decisions

## Commentary
### Why was this so challenging?
In a word: complexity. My original approach was to evaluate functions and programs across all of Arcane Finance, but the list quickly grew to >75 as its native programs were discovered to import other programs like `credits.aleo` and `token_registry.aleo`. 

Later, I discovered that Arcane has a special program called `arcn_compliance_v1.aleo` that is likely required for OFAC compliance in the United States, and which deanonymizes and makes public all private swaps and LP actions. It contains a consistent payload of inputs and outputs with the same 6 attributes:
1. Swap address (from)
2. Swap address (to; often equal to in)
3. Token 1 ID
4. Token 2 ID
5. Token 1 amount
6. Token 2 amount

_Should_ OFAC no longer sanction anonymity (see recent Tornado Cash ruling) this approach might no longer work. It's unclear if other protocols such as Pondo, Puzzle, etc. are OFAC-compliant or not. At this time, I haven't heard of regulatory action against any Aleo protocol.

### How does token registrations work?
This pipeline creates a new core table for token registrations, as all Aleo tokens use a generalized ID. Analysts can join to `dim__token_registrations` to learn when tokens were launched, decoded symbols and names, details about their decimal precision, and maximum supply. 

On Aleo, it's trivial to launch new tokens. As more are generated, they will land here and be available across protocol dashboards.

eric-laurello commented on models/silver/tokens/silver__token_registrations.sql:
```diff
@@ -0,0 +1,98 @@
+        block_id,
+        block_timestamp,
+        INPUTS
+    from aleo_dev.core.fact_transitions
```
change to ref

eric-laurello commented on models/silver/tokens/silver__token_registrations.yml:
```diff
@@ -0,0 +1,52 @@
+    
+    columns:
+      - name: tx_id_created
+        tests:
```
I don't think we should be testing the silver view, will be a lot of resources and we can just test gold

eric-laurello commented on models/silver/tokens/silver__token_registrations.sql:
```diff
@@ -0,0 +1,98 @@
+        block_timestamp,
+        value:id::string as id,
+        value:value::string as value,
+        row_number() over (partition by tx_id order by index) - 1 as input_index
```
why not use the index from the flatten?

eric-laurello commented on models/silver/tokens/silver__token_registrations.sql:
```diff
@@ -0,0 +1,98 @@
+
+parsed_inputs as (
+    select 
+        tx_id,
```
I typically do this in a slightly different way. It's produces the same output, but maybe it's slightly cleaner code?

`   select 
        tx_id,
        block_id,
        block_timestamp,
          OBJECT_AGG(
            input_index :: STRING,
            value :: variant
        ) AS j,
        J:"0"::STRING as token_id_raw,
        J:"1"::STRING as name_raw,
        J:"2"::STRING as symbol_raw,
        J:"3"::STRING as decimals_raw,
        J:"4"::STRING as max_supply_raw,
        J:"5"::STRING as external_auth_required_raw,
        J:"6"::STRING as external_auth_party
    from flattened_inputs
    where block_id != 186732
    group by tx_id, block_id, block_timestamp`

eric-laurello commented on models/silver/defi/silver__liquidity_pool_actions_arcane.sql:
```diff
@@ -0,0 +1,102 @@
+            ELSE 'OTHER'
+        END AS liquidity_action
+    FROM
+        {{ ref('core__fact_transitions') }}
```
add a top level cte that pulls from transitions with the incremental filter

eric-laurello commented on models/silver/defi/silver__liquidity_pool_actions_arcane.sql:
```diff
@@ -0,0 +1,102 @@
+    p.root_action,
+    p.liquidity_action,
+    p.liquidity_provider,
+    p.token1_amount / power(10, COALESCE(t1.decimals, 6)) AS token1_amount,
```
keep the unadj amount here too in case we're missing a decimal or the default 6 is not correct

eric-laurello commented on models/silver/defi/silver__swaps_arcane.sql:
```diff
@@ -0,0 +1,110 @@
+        tx_id,
+        program_id || '/' || function AS root_action
+    FROM
+        {{ ref('core__fact_transitions') }}
```
same comment as the lp model with th top level cte

eric-laurello commented on models/silver/defi/silver__swaps_arcane.sql:
```diff
@@ -0,0 +1,110 @@
+        REPLACE(inputs[4]:value, 'u128', '') AS amount_from,
+        REPLACE(inputs[5]:value, 'u128', '') AS amount_to
+    FROM
+        aleo.core.fact_transitions
```
ref

eric-laurello commented on models/gold/tests/tokens/test_core__token_registrations_recent.yml:
```diff
@@ -0,0 +1,75 @@
+      - dbt_utils.unique_combination_of_columns:
+          combination_of_columns:
+            - TX_ID_CREATED
+            - TOKEN_ID
```
we're joining on the token_id in other places. Should that be the uk?

eric-laurello commented on models/silver/defi/silver__swaps_arcane.sql:
```diff
@@ -0,0 +1,110 @@
+    unique_key = 'swaps_arcane_id',
+    incremental_strategy = 'merge',
+    merge_exclude_columns = ['inserted_timestamp'],
+    cluster_by = ['modified_timestamp::DATE', 'swapper', 'swap_from_name', 'swap_to_name'],
```
can just cluster by the mod_timestamp here

eric-laurello commented on models/gold/defi/defi__fact_swaps.sql:
```diff
@@ -0,0 +1,63 @@
+        tx_id,
+        succeeded,
+        swapper,
+        swap_from_amount,
```
naming wise we can drop the swap_ since it's a swap table. I also would rename swap protocol to just platform to be consistent

eric-laurello commented on models/gold/defi/defi__fact_swaps.sql:
```diff
@@ -0,0 +1,63 @@
+    meta = { 'database_tags':{ 'table':{ 'PURPOSE': 'SWAPS' }}},
+    unique_key = ['fact_swaps_id'],
+    incremental_predicates = ['DBT_INTERNAL_DEST.block_timestamp::DATE >= (select min(block_timestamp::DATE) from ' ~ generate_tmp_view_name(this) ~ ')'],
+    cluster_by = ['block_timestamp::DATE', 'succeeded',' swapper', 'swap_from_name', 'swap_from_id', 'swap_to_name', 'swap_to_id', 'root_action'],
```
can remove everything put the block timestamp from the cluster. We can put the others (maybe minute succeeded and root action into search optimization)

eric-laurello commented on models/gold/defi/defi__fact_liquidity_pool_actions.sql:
```diff
@@ -0,0 +1,66 @@
+    incremental_strategy = 'merge',
+    merge_exclude_columns = ['inserted_timestamp'],
+    cluster_by = ['block_timestamp::DATE'],
+    post_hook = "ALTER TABLE {{ this }} ADD SEARCH OPTIMIZATION ON EQUALITY(liquidity_pool_protocol,liquidity_provider,root_action,liquidity_action,token1_id,token2_id,token1_name,token2_name,liquidity_pool_protocol);",
```
base rule for SO is have a high level of cardinality on the column. I think it makes sense to cluster by the platform/protocol but not sure if we need the actions in there at all

eric-laurello commented on models/gold/core/core__fact_transitions.sql:
```diff
@@ -4,7 +4,7 @@
     incremental_strategy = 'merge',
     merge_exclude_columns = ['inserted_timestamp'],
-    cluster_by = ['block_timestamp::DATE'],
+    cluster_by = ['block_timestamp::DATE', 'program_id', 'function'],
```
maybe the same SO comment here, could be better in SO vs clustering

mrstepanovic replied:
I had to revisit the snowflake docs and you're right, analysts are likely to use range filters for dates and point lookups with equality predicates for programs and functions. Will review these accordingly

eric-laurello commented on models/gold/core/core__dim_token_registrations.sql:
```diff
@@ -0,0 +1,43 @@
+    unique_key = ['dim_token_registrations_id'],
+    incremental_strategy = 'merge',
+    merge_exclude_columns = ['inserted_timestamp'],
+    cluster_by = ['token_name'],
```
I would just leave the clustering off on this table

mrstepanovic commented:
@eric-laurello 

Alrighty, I should have addressed all of your comments with this commit. Thanks for explaining when to use SO vs. clustering and suggesting `OBJECT_AGG` instead of `CASE WHEN` in `silver__token_registrations`. I thought _not_ to include Aleo in this model because it's technically not a registered token, but it does simplify things a ton and would make queries simpler for analysts.

### Core Changes

- Simplified clustering and search optimization configurations across multiple models to focus on most commonly used columns
- Renamed columns in swap-related models to be more consistent:
  - swap_from_* → from_*
  - swap_to_* → to_*
  - swap_protocol → platform

### Token Registration Updates

- Added native Aleo token to silver__token_registrations with hardcoded values
- Simplified token registration parsing logic using OBJECT_AGG instead of multiple CASE statements
- Updated unique key from dim_token_registrations_id to token_id
- Modified tests to handle special case of Aleo token ID

### Performance Optimizations

- Reduced redundant CTEs in Arcane swap and liquidity pool models
- Added explicit dependencies for silver models
- Improved incremental logic by moving filters earlier in the query chain
- Added base CTE for transitions to reduce repeated logic / only query once

### Testing & Documentation

- Removed redundant token registration YML file
- Updated column descriptions to reflect renamed fields
- Modified unique combination tests to account for edge cases
- Added more specific null value tests with WHERE clauses (e.g. Aleo does not have a deployment `tx_id`)

eric-laurello commented on models/silver/defi/silver__swaps_arcane.sql:
```diff
@@ -0,0 +1,106 @@
+    a.swap_from AS from_id,
+    a.amount_to as to_amount_raw,
+    a.amount_to / power(10, t2.decimals) AS to_amount,
+    t2.token_name AS to_name,
```
sorry, just noticed we're using name here. We typically do symbol in the models. Is name better for Aleo? If not we should tick with symbol.

eric-laurello commented on models/silver/defi/silver__swaps_arcane.sql:
```diff
@@ -0,0 +1,106 @@
+    a.amount_from / power(10, t1.decimals) AS from_amount,
+    t1.token_name AS from_name,
+    a.swap_from AS from_id,
+    a.amount_to as to_amount_raw,
```
can you got with _unadj as the suffix - matches the standard we're going with going forward

eric-laurello reviewed (approved):
we're good!
